keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java

```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
         this.selectableValues = selectableValues;
     }
 
-    public T sample(int[] shape, KeanuRandom random) {
-        double sumOfProbabilities = getSumOfProbabilities();
-        double p = random.nextDouble();
-        double sum = 0;
-
-        if (sumOfProbabilities == 0.0) {
+    public GenericTensor<T> sample(int[] shape, KeanuRandom random) {
+        DoubleTensor sumOfProbabilities = getSumOfProbabilities(shape);
+        if (!sumOfProbabilities.lessThanOrEqual(0.).allFalse()) {
             throw new IllegalArgumentException("Cannot sample from a zero probability setup.");
         }
 
-        T value = null;
+        double p = random.nextDouble();
+        DoubleTensor sum = DoubleTensor.zeros(shape);
+        GenericTensor<T> value = new GenericTensor<>(shape);
+
         for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
-            sum += entry.getValue().scalar() / sumOfProbabilities;
-            if (p < sum) {
-                value = entry.getKey();
+            sum.plusInPlace(entry.getValue().div(sumOfProbabilities));
+
+            BooleanTensor mask = sum.greaterThan(p);
+            value.setWithMaskInPlace(mask.toDoubleMask(), entry.getKey());
+
+            if (mask.allTrue()) {
                 break;
             }
         }
-        if (value == null) {
-            T[] values = (T[]) selectableValues.keySet().toArray();
-            value = values[values.length - 1];
+
+        if (value.isNull()) {
+            value = new GenericTensor<>((T[]) selectableValues.keySet().toArray(), shape);
         }
 
         return value;
     }
 
-    public DoubleTensor logProb(T x) {
-        double sumOfProbabilities = getSumOfProbabilities();
-        if (sumOfProbabilities == 0.0) {
+    public DoubleTensor logProb(GenericTensor<T> x) {
+        DoubleTensor sumOfProbabilities = getSumOfProbabilities(x.getShape());
+        if (!sumOfProbabilities.lessThanOrEqual(0.).allFalse()) {
             throw new IllegalArgumentException("Cannot sample from a zero probability setup.");
         }
-        final double probability = selectableValues.get(x).scalar() / sumOfProbabilities;
-        return DoubleTensor.scalar(Math.log(probability));
+
+        DoubleTensor logProb = DoubleTensor.zeros(x.getShape());
+        for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
+            logProb.plusInPlace(x.equalsMask(entry.getKey()).timesInPlace(entry.getValue().div(sumOfProbabilities).logInPlace()));
+        }
+        return logProb;
     }
 
-    private double getSumOfProbabilities() {
-        double sumP = 0.0;
+    private DoubleTensor getSumOfProbabilities(int[] shape) {
+        DoubleTensor sumP = DoubleTensor.zeros(shape);
         for (DoubleTensor p : selectableValues.values()) {
-            sumP += p.scalar();
+            sumP.plusInPlace(p);
```

Comment:
Will the input shape always match the shape of the selectable values? If so, does it need to be an input? If not, should `sumP` be reassigned here as in some cases, when tensor shapes differ, a new tensor is created in `plusInPlace` (instead of reassigning in place)